src/Common/src/System/Net/SocketAddressPal.Unix.cs

```diff
@@ -10,156 +10,129 @@ namespace System.Net
 {
     internal static class SocketAddressPal
     {
-        public const int IPv6AddressSize = Interop.libc.sockaddr_in6.Size;
-        public const int IPv4AddressSize = Interop.libc.sockaddr_in.Size;
         public const int DataOffset = 0;
 
-        public static unsafe AddressFamily GetAddressFamily(byte[] buffer)
+        public readonly static int IPv6AddressSize;
+        public readonly static int IPv4AddressSize;
+
+        static SocketAddressPal()
         {
-            fixed (byte* rawAddress = buffer)
+            int ipv4AddressSize, ipv6AddressSize;
+            unsafe
             {
-                var sockaddr = (Interop.libc.sockaddr*)rawAddress;
-                switch (Interop.CheckedRead((byte*)sockaddr, buffer.Length, &sockaddr->sa_family))
-                {
-                    case Interop.libc.AF_UNSPEC:
-                        return AddressFamily.Unspecified;
+                Interop.Error err = Interop.Sys.GetIPSocketAddressSizes(&ipv4AddressSize, &ipv6AddressSize);
+                Debug.Assert(err == Interop.Error.SUCCESS);
+            }
 
-                    case Interop.libc.AF_UNIX:
-                        return AddressFamily.Unix;
+            IPv4AddressSize = ipv4AddressSize;
+            IPv6AddressSize = ipv6AddressSize;
+        }
```

Comment:
We should double-check the JIT's behavior with regards to treating readonly static Int32s as constants for non-beforefieldinit types.  I think it still will, but on the off-chance it doesn't, it'd probably be more efficient overall to have two static methods, each of which calls GetIPSocketAddressSizes, and use those in the declarations of the field to let the type be beforefieldinit.  Might be worthwhile to do that anyway for slightly more efficient access to the class in general.